python/src/aiconfig/eval/openai.py

```diff
@@ -0,0 +1,87 @@
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Protocol
+
+import lastmile_utils.lib.core.api as cu
+import openai
+import openai.types.chat as openai_types
+from aiconfig.eval import common
+from result import Err, Ok, Result
+
+
+@dataclass
```

Comment:
Good question. According to the zen of python, "There should be one-- and preferably only one --obvious way to do it." Therefore, the Python community created at least two complicatedly overlapping and non-obvious solutions to essentially the same problem, which is the fact that Python doesn't have an immutable record type.

Reply:
tl;dr I've been using immutable strict BaseModel (which is core_utils.Record) because it does auto-strict serialization and deserialization. However, that can be a blocker/PITA sometimes, which is when I go back to good old dataclass which essentially is just record with auto-constructor.